### ISSUE TITLE
Download task file system permissions

### ISSUE THREAD

**Issue by berkcebi** [Alamofire/Alamofire#2527]
Hey hey! 👋

Alamofire is using `URLSessionDownloadTask` for download tasks and the file created under the temporary directory by `URLSessionDownloadTask` has file system permissions set to `-rw-------`. (To be honest, I'm not sure why that is.) Alamofire then moves this file to the location provided by the user, keeping the permissions intact.

These default permissions are unlike most use cases, where they're set to `-rw-r--r--`:
- Downloading the file through the browser
- Downloading the file using a tool like `wget`, `curl`
- Downloading the file using AFNetworking
- Using Alamofire, requesting the file as a data object then writing to file

Either by default or when an option is provided, Alamofire could set the permissions up after moving the file from the temporary directory. I wanted to reach out and gather some feedback on this.

---

* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?

Used `DownloadTask` to download a file to a url, on macOS.

## What did you expect to happen?

The downloaded file to have the same file system permissions (by default), as downloading the same file through a browser, using a tool like `wget` or another networking library like AFNetworking: `-rw-r--r--`.

## What happened instead?

The downloaded file has a file system permission of `-rw-------`, not letting other users to access the file.

## Alamofire Environment

**Alamofire version:** 4.7.2
**Xcode version:** 9.3.1
**Swift version:** 4.1
**Platform(s) running Alamofire:** macOS
**macOS version running Xcode:** 10.13.4 (17E202)

**Comment by jshier** [comment]
Thanks for the report! This is an interesting difference that doesn't make much sense. AFNetworking moves it's file like this:
```objc
[[NSFileManager defaultManager] moveItemAtURL:location toURL:self.downloadFileURL error:&fileManagerError]
```
Alamofire moves it like this: 
```swift
try FileManager.default.moveItem(at: url, to: destinationURL)
```
As far as I can tell, there shouldn't be any difference there. And neither library set any additional properties on the file, like special POSIX permissions. 

We could easily add said permissions, but I'd rather figure out why this is happening first, and if it's something that does need to change, integrate a way for users to control the behavior rather than just having Alamofire set the permissions directly.

Also, you wouldn't happen to have a quick demo app to replicate this, would you? It's been a while since I've built a Mac app.

**Comment by berkcebi** [comment]
Right, moving doesn't cause the issue actually. Basically the temporary file created by `URLSessionDownloadTask` has these permissions by default.

I just realized that AFNetworking uses `URLSessionDownloadTask` as well. So the results are the same for Alamofire and AFNetworking indeed—sorry for the confusion there.

Honestly, I'm not sure why the file created by `URLSessionDownloadTask` has these permissions though. Instead of using a download task, if I make a request to the same url and write the response data to file manually, permissions are set to `-rw-r--r--`.

I've demonstrated that in the sample project here: https://github.com/berkcebi/down

**Comment by jshier** [comment]
I'm guessing this happens because it's the safe default for file downloads, plus most users never need to make files available to there users. 

That being said, since the file URL is available in the `DownloadResponse`, what would there be for Alamofire to do here? Users can easily change permissions in their response handlers, so I'm not sure there's a better solution here. I'm reluctant to do anything automatically since Apple's behavior seems to be the safe default. So what would you want here?

**Comment by berkcebi** [comment]
I definitely see your point on sticking with Apple's behaviour, if it is indeed a safe default for downloads. In that case, wouldn't you expect to have the same permissions for files downloaded through say, Safari though?

As stated on [`URLSessionDownloadTask` documentation](https://developer.apple.com/documentation/foundation/urlsessiondownloadtask), the file created is placed in a temporary directory and should be moved or consumed right away when the task is complete. The fact that it's a temporary file might also be the reason why the permissions are set this way actually.

Alamofire offers a quick way to move the temporary file via `DownloadFileDestination`. While moving the file, if a common practice is to update the file permissions as well, this might be added as an option to `DownloadFileDestination` as well. Does this make sense?

**Comment by jshier** [comment]
I'll add this as a feature request, though it's unlikely we'll get to it for the initial Alamofire 5 release.

**Comment by berkcebi** [comment]
Thanks @jshier! I'd be happy to contribute actually. If this makes sense to you as well, I can start working on it.

**Comment by mazz** [comment]
Same issue on iOS, Alamofire 4.8.x

For those of you who cannot wait for an Alamofire fix:

```
                            do {
                                // need to manually set file perms: https://github.com/Alamofire/Alamofire/issues/2527
                                try FileManager.default.setAttributes([FileAttributeKey.posixPermissions: NSNumber(value: 0o644)], ofItemAtPath: internalFileDownload.localUrl.path)
                            } catch {
                                print("error while setting file permissions")
                            }
```
